### PR TITLE
refactor star file reading

### PR DESCRIPTION
**Pull request by alisterburt**
Here's an attempt at refactoring star file IO into something a little more flexible, keen to hear your thoughts

The main ideas behind this implementation are that:
- main function `read_star` reads raw data using starfile then dispatches reading
- reading is attempted by explicit "reader_functions" sequentially

I think that overall this is more explicit than the use of conventions as we were working with before which quickly becomes an unreadable nested if/else mess where interpretation is mixed with parsing. The price is a little bit of code duplication, although some refactoring could be done

relion 3.1 style shifts are now handled inline, optics and particles dataframes are merged so that shifts can be calculated properly - the implementation is robust to the possibility of multiple optics groups (which can have different pixel sizes, happens in single particle sometimes)

One thing that's lost in this implementation is the ability to use your regex wizardry for filename matching, I'm sure it can be added back in I just didn't understand it enough to do so

### CODE REVIEW

**Comment by alisterburt** [comment]
this will close #59 

should also note that tests should be added - I'll try to add the test data repo with the following tomorrow
- a dynamo table file
- a relion 3.0 star file
- a relion 3.1 star file

File: peepingtom/io_/read/star/utils.py [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+        name = Path(micrograph_name).stem
+
+        coords = df_[coord_headings]
+        shifts_px = df_[shift_headings] / df[pixel_size_heading]
+        xyz = (coords + shifts_px).to_numpy()
+
+        eulers = df_[euler_headings]
+        rotation_matrices = euler2matrix_rln(eulers)
+
+        properties = {key: df[key].to_numpy() for key in df.columns}
```
Comment (alisterburt) [inline peepingtom/io_/read/star/utils.py:54]: should be df_[key]

File: peepingtom/io_/read/star/star.py [hunk 2]
```diff
@@ -1,82 +1,18 @@
...
+    raw_data = starfile.read(star_path, always_dict=True)
+
+    failed_reader_functions = []
+    for style, reader_function in reader_functions.items():
+        try:
+            particle_blocks = reader_function(raw_data)
+            return particle_blocks
+        except ValueError:
+            failed_reader_functions.append((style, reader_function))
+    raise NotImplementedError(f'Failed to parse {star_path} using {failed_reader_functions}')
```
Comment (brisvag) [inline peepingtom/io_/read/star/star.py:18]: I don't think is necessarily a `NotImplementedError`. The file may be corrupt, for example. Using the more generic `ValueError` allows us to catch this higher up together with the rest of IO.

**Review by brisvag — CHANGES_REQUESTED** [review]
I don't like this much code duplication, especially considering how the same stuff was basically written a third time in #62.

However, I really like the sub-dispatcher system. Doing something similar to what I did higher up for file extensions looks clean and makes sense.

I think maybe we should find another way of "splitting" the IO, because doing it by extension like this forces us to split things that behave effectively identically like tbl and several types of starfiles.
Maybe splitting by type of DataBlock generated? This way we could have a common `utils` function for similar things.

Either way, I still think the `conventions` dict should exist in some shape or form. Not only it makes it easier for people to add their own specific readers, so long as they are star-like (if not, wrong part of the codebase), but it allows us to maintain better and more easily add new versions and whatnot.

**Comment by alisterburt** [comment]
@brisvag just summarising our discussion a bit here:

- we agreed that star file and dynamo table handling should not be handled by the same Handler
- dispatching data already read from disk is the right way to go within the star reading module, despite the nested-ness
- maybe handling star IO more generically is useful, given the duplication we see here

my feeling is still that I like these explicit handlings of specific cases despite the presence of common logic, it gives the function single responsibility and is nice and explicit

I think this is now ready to merge!

**Review by brisvag — APPROVED** [review]
Agreed, we may come back to the duplication in the future, but for now it's good to go!

**Pull request merged**